

Nanopore/Canu assembly tutorial

The data for this week have been generated by my research lab, in particular my graduate student John Sutton. Copy the data into your home directories on ASC. They are in the shared directory. Remember that in order to use them you need to copy them into your home directory and work there.

$ cd shared
$ cp *YourGroup.fastq* ~

First, do some sanity checks on the reads. How many reads are there?

$ wc –l *YourGroup.fastq* # 

First, load the Canu module

    $ module load canu

Canu should be accessible, test this and see if you get a program description

    $ canu

You can read more about the canu assembly software here https://canu.readthedocs.io/en/latest/quick-start.html

The Canu pipeline uses a three-step procedure for assembly with sequence reads corrected, trimmed and assembled. The default is to run all three steps and this is what we will do. You will need to use vi to write a Canu assembly script. Type

    $ vi canu_assembler.sh

Hit 'i' for insertion mode and type your script in

    #!/bin/bash

    source /opt/asn/etc/asn-bash-profiles-special/modules.sh

    module load canu

    # place commands here

    canu -useGrid=false -p nanopore -d Assembly -genomeSize=5M stopOnReadQuality=false -nanopore-raw *YourGroup.fastq*

Save your script by typing ‘escape’ then :wq. Make it executable by typing

    $ chmod +x canu_assembler.sh

Submit your script to the class queue

    $ run_script canu_assembler.sh

You will need a LOT of resources to assemble so maximize the number of processors and the memory available in the class queue. Also know that these assemblies take a while (one week is not uncommon!) so get started ahead of time.

I have included an accessory script, count_fasta.pl that can be used to produce a histogram-appropriate file of sequence lengths and some summary stats. You call it like this:

    $ perl count_fasta.pl –i 10000 ./Assembly/contigs.fa

Where the ‘-i’ flag sets the bin size for the histogram and

    ./Assembly/contigs.fa 

is the path to your assembled contigs file. Try running it with –i 100 , -i 1000 and other values to see how your k-mer size affects the distribution of assembled sequences in your final output.

Now we need to figure out what this organism is. The easiest way to do this when you have a large contiguous sequence is to copy a section of the assembled sequence and paste it into the NCBI Blast browser.

Your homework is to write a report describing your assembly experience and answering these specific questions:

    What is your organism?
    What are the final statistics for your assembly (Total size, N50, number of scaffolds)?

You must print your report and hand it in. You may either drop it by my office (SEC 2328) or the Biological Sciences office (SEC 1325). **You will also give a short (10 minute) presentation to the class about your assembly experience and your organism**. 
