**Session 3: Working with Files**

Last time we learned how to navigate a directory structure using shell commands. This session will focus on using shell commands to manipulate files.

**Getting the file**

First, log on to the ASC.

    $ ssh -l yourusername here dmc.asc.edu

You should see an occupant in your home directory, a directory called **shared**. We will use this directory for sharing data files. 

We are analyzing a file of DNA sequence from a recent Illumina HiSeq run (this is only part of the file). Sequencing files are very large so we compress them to save space. The file should be in the **shared** directory in your home directory. You can read things in **shared** but you can’t edit so copy it from **shared** into your home directory. Navigate into the **shared** directory

    $ cd ./shared

and copy the file into your home directory. Some other standard shell commands are cp (copy) and mv (move). Both commands take the syntax **operation - from location 1- to location 2**. 

    $ cp ./file.fastq.gz ~

Ordinarily we could copy with the relative location reference but the **shared** directory is a symbolic link and we need to copy files out of it with an absolute path. Do you remember what the ‘.’ and ‘..’ location references indicate? What is the difference between a relative and absolute path?

A symbolic link is a secondary way to access a directory. It means the **shared** directory does not actually live inside your home directory, it lives somewhere else and the ASC folks have added a symbolic link to let you access the data. For example, try

    $ cp ./file.fastq.gz ../

What happens? Now try

    $ pwd

and try to copy the file with the *absolute* path to your home directory

    $ cp ./file.fastq.gz *yourpathhere*

What happens?

Now we are ready to start learning how to work with files. Navigate back to your home directory

    $ cd ../

or

    $ cd ~

and check the file size with 

    $ du -h file.fastq.gz

What does "du" stand for? What is the -h? Try

    $ man du

and remember to use ‘q’ when you are ready to quit looking at the manual.

First we need to uncompress or unzip the file.

    $ gunzip file.fastq.gz

What is “gunzip”? Use

    $ man gunzip

to examine the manual. You should see the name of the file change from having a '.fastq.gz' extension to '.fastq'

    $ ls 

Check the file size again

    $ du -h file.fastq

Is it different from before? How much of a difference in size does the compression account for?

Look inside the file with

    $ more file.fastq

In order to quit "more," type 

    $ q

and you should return to the prompt. You can also use *less*, *head*, and *tail*. Play around with these a bit, for example

    $ less file.fastq

Hint: look at the bottom of your screen when you try ‘more’ or ‘less’.

Then try

    $ head file.fastq

    $ head -1 file.fastq

    $ head -10 file.fastq

    $ tail file.fastq

    $ tail -2 file.fastq

This is a *fastq* file. It has four lines for each DNA sequencing read which are 1) the sequence read identifier; 2) the DNA sequence (or read); 3) "+" and 4) Quality scores for each base. We will go into more depth about .fastq files later but for now we will use this file to learn more shell commands. We typically use '.fastq' - note the 'dot' in the front there- to denote that this is a file type.

Count the number of unique lines in the file with 'wc'

    $ wc file.fastq

What is the output? (Note: this is a real file and counting lines will take a little while) 

Try 

    $ man wc

to understand the columns. Now try 

    $ wc -l file.fastq

where wc -l = "word count by line." How many lines do you have? How many actual DNA sequences do you have?

Again, wc is a very powerful command, so take some time to understand it. Type 

    $ man wc

and read through the manual pages. 

How can you use wc to find the sequence read length (tell how many characters are in each DNA sequence)? Hint: it is easiest if you combine wc with another command. Double hint: there are always many different ways to do things in Unix and there are many correct answers to this question.

For example, my solution would be:

    $ head -2 file.fastq | tail -1 | wc -c

What is this doing? Try to go through each command individually and understand the pieces of the operation. For example, start with

    $ head -2 file.fastq

then

    $ head -2 file.fastq | tail -1

then

    $ head -2 file.fastq | tail -1 | wc -c

We are going to explore the standard shell commands cp (copy) and mv (move). Both commands take the syntax **operation - from location 1- to location 2**. To explore these, make a new directory

    $ mkdir ./that_directory

and move your file to it

    $ mv file.fastq ./that_directory

Now type 

    $ ls

What happens?

Navigate to your new directory and try again

    $ cd that_directory

    $ ls

Did you find file.fastq?

You can also create a copy of your file with a different name

    $ cp file.fastq file_copy.fastq

or with the same name but in a different directory

    $ cp file.fastq ../

What happens?

**Streams, pipes and redirection**

First we will make a small file to operate on before doing analyses on our large file.

Type

    $ head -10000 file.fastq > small_file.fastq

You have now created a small file for testing. Count the files in your original file.fastq and your new small_file.fastq like this

    $ wc -l file.fastq
    
    $ wc -l small_file.fastq

Type

    $ cat small_file.fastq

What happens? The "cat" command will concatenate files. Here, we only give it one file and so it concatenates that file with our standard output (stdout), which is our terminal. If we give it two files we can stick them together, end to end. Try

$ cat small_file.fastq small_file.fastq > new.fastq

What happens (Hint: Check how many lines are in new.fastq with)
    
    $ wc -l new.fastq

Vertical concatenation happens with "cat" and horizontal concatenation happens with "paste." For example, we often want to use data in a column format. Type

    $ cat small_file.fastq | paste - - - - > column.fastq

Look at your file

    $ less column.fastq

What happened?

"cat" is concatenating your file.fastq to the terminal (stdout), and the "|" character pipes the output of one command to another, here from "cat" to "paste." This is really one of the most powerful things about UNIX, the ability to both pipe the output of one command into another, and simple file redirection. Here, the redirect puts the stdout of paste into a file called column.fastq. Look at the file to see how it is formatted. 

We can easily manipulate the columns of this file with the "cut" command. Try

    $ cut -f 1 column.fastq

Now try 

    $ man cut

What is the "-f" doing? What happens if you do 2 or 3? Can you combine them (for example, 1-2, 1-3)?

Try piping multiple commands

    $ cat column.fastq | cut -f 1 | sort

The Illumina read identifier is a physical location in the machine that tells us:

sequencing machine:run(numerical):run(identifier):lane:tile:x-coordinate:y-coordinate

Is your data from one tile or multiple tiles?

A very useful command is ‘uniq’ which stands for unique. You can use it to streamline multiples down to a single instance or to count instances with the -c flag.

One of the big problems in bioinformatics is that different people work on data generation and analysis and the molecular people know little about what the bioinformatics people are doing and vice versa. For example, a common molecular protocol is to add a short tag or ‘barcode’ to a molecular sample so you can sequence multiple samples together. We can use these barcodes after sequencing to separate samples. The barcode is typically the first 5-7 bp of the sequencing read. We can check for barcodes like this:

    $ cat small_file.fastq | paste - - - - | cut -f 2 | cut -c1-6 | sort | uniq -c | sort -nr | head -20

Take some time to build this command piece by piece an play around with the flags (options) so you can really understand what’s going on here.

Do you think this sample has barcodes? How many? What are they? In addition to count, what other evidence do you see that these are barcodes and not just random sequence?

Try this with the large file.fastq. It will take some time to get through the large file.

In-class credit:

1) List the possible barcodes and counts for each from your analysis of file.fastq

2) List the new commands you have learned this session and give a short (2-5 word) description of each. 

