**BLAST on ASC**

Now that we know about BLAST and the Alabama Supercomputer, we will learn how to use BLAST on the Alabama Supercomputer. The NCBI Blast web interfaces (and other web interfaces) are convenient for single sequence queries but if we have many queries we want to deal with it quickly becomes tedious. Using BLAST on the ASC allows you to scale up your work and easily format the output. 

The sequences we are working with today are a set of miRNA sequences obtained from the nematode worm *Caenorhabdtis latens*. miRNAs are short noncoding RNA (typically 20-25 nucleotides) that regulate gene expression by targeting mRNA for degradation or by blocking translation. We know that some miRNAs are highly conserved across evolutionary scales (for example, the mature sequence of *let-7* is perfectly conserved from nematodes through humans) while others show rapid rates of change. We also know that mutations in certain miRNAs are associated with human diseases and cancer progression. Right now scientists study miRNAs through discovery-based research; that is, we target each individually and identify what happens to the organism when we mutate it or knock it down. In order to gain a comprehensive understanding of the functional effects of miRNAs we need a better understanding of the evolutionary constraints acting on miRNAs. For example, can a miRNA tolerate mutations in any region or are only certain mutations functionally permissible? 

Part of my research is studying how miRNAs vary within populations and linking this variation to larger evolutionary patterns of change and conservation. I have a set of miRNA sequences that were obtained by sequencing small RNAs expressed in the nematode worm *C. latens*. *C. latens* is a distant relative of the model nematode *C. elegans.* In order to identify the nearest relatives of each miRNA and the sequence-level changes that have occurred between the relatives we will use BLAST to identify orthologous (i.e., descended from the same ancestral sequence) miRNAs between *C. latens* and *C. elegans.*

The first step is obtaining the set of *C. latens* miRNAs. Log on to your account on the Alabama Supercomputer

    $ ssh -l yourusernamehere dmc.asc.edu
    
Navigate into the *class_shared* directory
  
    $ cd class_shared

You should see the file Clatens\_miRNA.fasta in the *shared* directory. Copy this file into your home directory.

    $ cp Clatens_miRNA.fasta ~

First, do some sanity checks on the file. Look at the file format

    $ more Clatens_miRNA.fasta

How many lines are there? 

    $ wc -l Clatens_miRNA.fasta

How many individual miRNAs is that? 

If you read the names of each miRNA they start with the fasta identifier ‘>’ followed by the species abbreviation (cla for **C. la**tens), mir to indicate it is a miRNA, the number of the gene and a ‘5p’ or ‘3p.’ miRNA are formed from a longer segment that is folded into a ‘hairpin’ structure and the loop is cleaved off. The resulting double-stranded molecule has one 5p sequence and a complementary 3p sequence. People used to think that the 5p sequence was dominant and the 3p less so but with additional studies we have found that is not the case and both forms are often expressed. We only have mature sequence for those miRNAs that have actually been experimentally documented so some miRNAs have 3p and 5p sequences while others only have one sequence. 

In order to identify evolutionary relationships we will need to use BLAST to align these miRNAs to known miRNAs in other species. Copy the first one (it should be *let-7*) and go to the web NCBI BLAST page. Enter the sequence in the query and see what you get. We know our molecule is a miRNA, are you getting a related miRNA as a top hit? Change the databases you are searching against. Can you get a related miRNA as a top hit with any of the databases? Why do you think you are having problems finding related miRNAs with BLAST and the NCBI databases? (Hint: tiny)

Because of their significance in human health and disease miRNAs have their own website, miRBase (www.miRBase.org). Navigate there now with your web brower. miRBase hosts all known miRNA sequences and has several functions. The ‘Search’ page specifically searches miRNA databases with specially tuned BLAST or another program called Ssearch. You can also browse miRNAs by organism, download sets of miRNAs, and submit sequences if you are analyzing novel miRNAs. 

Navigate to the ‘Search’ page and paste your *let-7* sequence into the box. Search mature miRNAs with BLASTn and leave the e-value and maximum number of hits as default (10 and 100). You should get a list of >100 *let-7* sequences in different organisms. The results include a ‘query start,’ ‘query end,’ ‘subject start,’ and ‘subject end’ for each hit. Look at the ones at the top of the page and then further down. The score and e-value tell you about the quality of your alignment. How do these compare to the values you obtained with previous BLAST searches (hint: with longer molecules)? 

Click on ‘align’ at the end of the first column. You should see a blast alignment similar to those on the NCBI BLAST page. Now click on the name of your top hit, *cel-let-7-5p*, and it will take you to another page. Here you should see information about the organism, the sequence, the position within the full miRNA stem-loop, the experimental evidence for the miRNA, the chromosome coordinates, and any information available about how the miRNA was discovered and putative functions. The miRNAs *let-7* and *lin-4* were discovered prior to the standardized mir-XXXX naming system was developed and this is why their names are different. 

We want to assign a *C. elegans* relative to every *C. latens* miRNA (if possible; some will not have relatives). It would be tedious and boring to do a manual “brute force” search like this on miRBase. Instead, we are going to use BLAST on ASC to do the work for us.

Go to the ‘Download’ page, right-click on the ‘mature.fa’ file and click ‘Copy link location.’ Now go back to your ASC terminal (make sure you are in your directory) and type

    $ wget ***PASTE COPIED LINK IN HERE***

Don’t actually type the part above in bold; paste the copied link in. You should have something like

    $ wget ftp://mirbase.org/pub/mirbase/CURRENT/mature.fa.gz

wget and curl are two handy Unix programs that can help us download data from public sites. Once you hit return you should connect to the miRBase ftp site and download the file. It is zipped, so unzip it with 

    $ gunzip *gz

Memory jog - using our wildcard \* character here. 

Take a look at the file and do some sanity checks. Use disk usage to check how large it is

    $ du -h mature.fa

and check how many lines and miRNA sequences you have. 

    $ wc -l mature.fa
    
    $ grep -c ">" mature.fa

Here we are asking grep to count (-c) every instance of the .fasta header '>'. We can also get more specific and ask for lines that start with '>' with a '^' character. For regular expressions '^' denotes start of line and '$' denotes end of line. For example, try

    $ grep -c "^>" mature.fa
   
and 

    $ grep -c ">$" mature.fa
    
In the first command you are counting how many lines start with '>' and in the second command you are counting how many lines end with '>'. 

We are going to make our own set of *C. elegans* miRNAs to search against. We do this by extracting all *C. elegans* miRNAs with grep. Identify what the options are doing through the grep manual. For example, do you remember what the -A 1 means? What about the -v "^--"?

    $ grep -A 1 "Caenorhabditis elegans" mature.fa | grep -v "^--" > Celegans_miRNA.fasta

We can now use our *C. elegans* file to create our own BLAST database. Before we set up our script it is helpful to look at a few parameters so we know what we are setting up. Load the blast module

    $ module load blast+

and type

    $ module list

to make sure it loads. Type 

    $ blastn — -help

to see a list of the parameters blast requires and/or accepts. Unix BLAST has high functionality but limited help. For example, we need to make a blast database with ‘makeblastdb’ but we can not actually get UNIX BLAST to give us these kind of instructions in the Unix system. 

Why do we want to make our own database? First, as you have already seen the default BLAST databases don’t work well when we have very specific queries or are working with short sequences like small RNAs. Second, it will be much faster to run BLAST if our database is smaller (less to search) and the database is located in the same location as our working files (no input/output bottlenecks). 

In order to see the parameters required and/or accepted in making a BLAST database type

    $ makeblastdb

This is another inconsistent feature. Type

    $ blastn

What happens? Type

    $ makeblastdb —help

What happens?

Try 

    $ makeblastdb -h

What happens?

We have to access the help functions with different syntax for different modules of the same program. Because of this you have to have an arsenal of possible help calls, including 1) calling the program with no options; 2) calling the program with an —-h option; 3) calling the program with a -h option; 4) calling the program with a -help option or --help. 

Type 

    $ makeblastdb -in Celegans_miRNA.fasta -dbtype nucl -out Celegans

You should see a series of files with Celegans.nxx names produced. These are the index files BLAST uses to accelerate database searching. Note that we were able to do this on the head node because it is a small file and runs quickly; we need to run the BLAST search through a script. Before we make our script look at the blastn options again. The most significant are ‘db’ for which database to search and ‘query’ for which file we want to search with.  

Use vi to create a new script

    $ vi blast_script.sh

and insert the following text:

    #!/bin/bash
    source /opt/asn/etc/asn-bash-profiles/modules.sh

    module load blast+
    blastn -db Celegans -query Clatens_miRNA.fasta > blast_output.txt

Save your script by hitting ‘escape’ and then ‘:wq’ to exit vi. Make the script executable:

    $ chmod +x blast_script.sh
    
and submit the script:

    $ run_script blast_script.sh

then go through the prompts. You want to use

(1)	the class queue

(2)	1 processor core (the smallest amount of resources, so you get going the fastest)

(3)	A one hour time default (01:00:00)

(4)	64gb of memory

(5)	any queue (hit return)

What happens?

Command-line blast has a parameter, ‘word\_size’ that is not optimized for short sequences. Edit your script so the blastn line looks like this 

    blastn -db Celegans -word_size 11 -query Clatens_miRNA.fasta > blast_output.txt

and run the script again. What happens now?

It is sometimes easier to see what’s going on with a tabular output. Edit your script so the Blast line looks like this and run again:                         

    blastn -db Celegans -word_size 11 -outfmt 6 -query Clatens_miRNA.fasta > blast_output.txt

What happens now? The columns tell you:

1) Query

2) Blast hit

3) Percent match

4) Length of aligned sequence

5) Number of polymorphisms (mismatches)

6) Gap score 

7) Query start 

8) Query end 

9) Subject start

10) Subject end

11) e-value

12) bitscore

What do you see in the results? Make sure you are running with and without the tabular format (-outfmt 6) so you can visualize the alignments as well as the matches.

*C. elegans* and *C. latens* are separated by \~ 40 million years of evolution. As a final step we will use Clustal (https://www.ebi.ac.uk/Tools/msa/clustalo/) to align a set of miRNAs and visualize their relationships. miRNAs are assigned to ‘families’ based on their similarity and miRNAs that are very similar but not exactly the same are indicated with a letter after the numerical part of their identifier. For example, the mir-35 family is large and appears to have evolved through multiple tandem duplications. 

Extract all of the *C. latens* and *C. elegans* members of the mir-35 family like this 

    $ grep -A 1 'mir-35[a-z]' Clatens_miRNA.fasta

    $ grep -A 1 'miR-35-' Celegans_miRNA.fasta

and align them with Clustal-Omega on the EMBL-EBI web site. You should be able to visualize both the multiple alignment in a sequence format and a ‘phylogenetic tree’ of the sequences. In order to do this rigorously you need to do 2 alignments: 1) all *C. latens* mir-35 sequences; and 2) all mir-35 sequences from the two species. Which *C. latens* mir-35 is the most similar to the *C. elegans* mir-35?

Questions for homework:

1) How many miRNA are in your file?

2) Which NCBI databases align your *let-7* sequence to a miRNA in another species?

3) Why is it difficult to find related miRNAs with BLAST and the NCBI databases?

4) What do the grep flags -A and -v do?

5) What are some of the problems with the default BLAST output? 

6) How many BLAST hits do you get between *C.latens*-*C.elegans* and what is the average level of sequence identity (percent match)? Hint: every line is a hit and you can count lines with 

        $ wc -l blast_output.txt

If you were looking at just your *C.latens*-*C.elegans* BLAST results, what conclusions would you make about how miRNAs evolve? For example, is there a high level of sequence identity (slow evolution) or low (fast evolution)?

7) Which *C. latens* mir-35 is the most similar to the *C. elegans* mir-35? What can you infer about the evolutionary relationships between the mir-35 miRNAs in *C. latens*? 

8) You have gone through a complete protocol or ‘workflow’ for identifying orthologous miRNAs in related species. Create a diagram that illustrates this workflow. Include major steps and arrows indicating movement between steps. Your workflow should be detailed enough that someone else with a bioinformatics background could implement your protocol.

