Noninvasive Prenatal Diagnosis

For this exercise we will analyze data from the NCBI sequence read archive and perform a noninvasive fetal aneuploidy detection. BGI (formerly the Beijing Genomics Institute) submitted samples to the US NCBI sequence read archive that were used in designing their NIFTY test. The test is described on their website, www.niftytest.com. Read the website, particularly the portion for ‘healthcare providers.’

The basic workflow is outlined below, and specific instructions follow that. In this exercise I will give you some specific instructions on commands but you will also need to work with your group to figure out the commands based on the workflow and instructions and the commands we have learned in previous classes. You will also need to read the documentation for the programs that we will use and figure out the commands based on your knowledge of UNIX. Previous groups have gotten stuck on a few steps so we will do a 'cooking class' implementation where I will provide some prepared files for you at different steps.

Workflow:

1) Retrieve sample from the NCBI ftp site
2) Use the sra toolkit to convert the .sra binary file to a workable .fastq file
3) Inspect the .fastq file
4) *Cooking class* Use bowtie2 to align the .fastq file to the human genome reference sequence in ascii format (.sam)
5) *Cooking class* Use samtools to create a sorted binary alignment file (.bam)
6) *Cooking class* Use samtools to create an index for your file (.bai)
7) *Cooking class* Use bedtools to create a genome coverage file
8) Use awk to calculate coverage means and standard deviations for chromosomes 

The first thing we will do is download the data from the NCBI sequence read archive. The data are described here: http://sra.dnanexus.com/studies/SRP037560/experiments and the study described here: http://journals.plos.org/plosone/article?id=10.1371/journal.pone.0092192 . We will jointly analyze one sample - 

All groups - SRR1167207

The easiest way to download them is to retrieve the file from the NCBI site. You format the address like so depending on your sample:

https://trace.ncbi.nlm.nih.gov/Traces/sra/?run=SRR1167210

Navigate to "Data access," right click on the file under "Name" and download it to ASC with wget

	$ wget **paste in here**

Once you have the sample on your Alabama Supercomputer account you will need to transfer it from .sra (a binary NCBI-only file format) to a workable fastq file. NCBI provides their own sra toolkit, it is available on ASC as the module ‘sra.’ The documentation for converting .sra to .fastq is here http://www.ncbi.nlm.nih.gov/Traces/sra/sra.cgi?view=toolkit_doc&f=fastq-dump.

An example script:

Start the text editor with 

	$ vi sra2fastq.sh
	
Then type your script:

	#!/bin/bash

	source /opt/asn/etc/asn-bash-profiles-special/modules.sh

	module load sra

	# place commands here

	for file in *sra

	do

		fastq-dump --split-files ${file}

    	done

Reminders for running scripts:

1) make it executable with 
	
		$ chmod +x sra2fastq.sh

2) ASC assumes that everything is located in your home directory. If you are working in another directory (if you, for example, made a “prenatal” directory and everything is in there) you must specify the path to your files in your script. For example, ~/prenatal/*.sra .

3) Submit your script with 

		$ run_script sra2fastq.sh

Once we have the data in .fastq format we can use a short read alignment program like Bowtie2 (http://bowtie-bio.sourceforge.net/bowtie2/index.shtml ) to align the reads to a reference human genome sequence. 

Reference genomes are large and commonly used reference sequences are located in /opt/asn/bio/unzipped/bowtie2/ . Alignment programs require that reference genomes are indexed with compression algorithms that facilitate rapid searching. I wrote a script below that performs a sample alignment. You will need to change it for your sample numbers.

*** COOKING CLASS STARTS HERE ***

Start the text editor to create a script

	$ vi prenatal.sh

And use the following script for alignment. Be sure to change the SRR number to your data. 

	#!/bin/bash

	source /opt/asn/etc/asn-bash-profiles-special/modules.sh

	module load bowtie2

	# place commands here

	bowtie2 -x /apps/bio/unzipped/bowtie2/Homo_sapiens/UCSC/hg38/Sequence/Bowtie2Index/genome -1 SRR953066_1.fastq -2 SRR953066_2.fastq -S bowtie2.sam

.sam is a great format for inspecting your reads and getting to know your data. The file starts with a number of @SQ lines where every line corresponds to a section of the reference genome. The SN: field gives the chromosome and the LN: field gives the length. 

Take a look at the actual .sam alignment lines ( Hint: grep for the lines that don’t begin with @, i.e. -v, and look at the first 10 lines *head*. Also look at each column of data individually with cut -f 1 ; cut -f 2 ; etc). You can find information about the .sam format here  https://samtools.github.io/hts-specs/SAMv1.pdf and here http://genome.sph.umich.edu/wiki/SAM .

In order to format our .sam file and analyze genome coverage we need to convert it to a binary format. Our binary file must also be sorted and indexed. We use samtools for this (http://www.htslib.org/doc/samtools.html). Use the example script below but replace ‘file.sam’ and ‘file_sorted’ with the names of each of your files. 

Start the editor with

	$ vi sam2bam.sh

And use this script for sam-to-bam conversion

	#!/bin/bash

	source /opt/asn/etc/asn-bash-profiles-special/modules.sh

	module load samtools
	module load picard/1.79

	# place commands here

	samtools view -bS bowtie2.sam > bowtie2.bam 

	samtools sort bowtie2.bam > bowtie2_sorted.bam

	java -jar /opt/asn/apps/picard_1.79/picard-tools-1.79/MarkDuplicates.jar INPUT=bowtie2_sorted.bam OUTPUT=bowtie2_dedup.bam METRICS_FILE=metrics.txt REMOVE_DUPLICATES=true

The last line uses a software suite called picardtools to remove duplicate sequencing reads in our file. This is how we select Unique Reads for analysis. Next time we will analyze coverage across the human genome. 

Part II

Today we will use use bedtools ( http://bedtools.readthedocs.org/en/latest/ ) to analyze the sequencing coverage across the genome. Bedtools performs genome arithmetic. In particular the genomecov tool will calculate genome coverage 
( http://bedtools.readthedocs.org/en/latest/content/tools/genomecov.html ). ASC has a bedtools module. Note that we are using a .bam file as input ( Hint: the -ibam option ). You can try a few different ways to calculate the genome coverage to compare the output. Below is my default script (again, you will need to replace the file names with your file names):

Start vi

	$ vi genomecoverage.sh

Type 'i' to insert and type the following script

	#!/bin/bash

	source /opt/asn/etc/asn-bash-profiles-special/modules.sh

	module load bedtools

	# place commands here

	bedtools genomecov -ibam bowtie2_dedup.bam -d > genome_coverage.txt

Hit escape and type ':wq' to save and exit. 

Bedtools has two different genome coverage options. If we use ‘-d’ we can output the coverage for every position in the human genome but we get a very large file. If we use ‘-bg’ we can collapse areas of similar coverage together. Calculations done on this file are not as accurate but they can save space and time. Here, we will use -d because of our low-coverage data. 

This step will take a while to complete because it has to index and check every position in the 3Gb+ human genome. Check if you are still running with 

	$ squeue

*** COOKING CLASS ENDS HERE ***

** I will provide these files for you to analyze with your group **

Once you have a file with genome coverage we can use awk to calculate the sequencing coverage across each chromosome. In order to diagnose the subject we will use both numerical and graphical output. We can calculate chromosome mean coverage and standard deviation in coverage with awk. There is an awk statement to calculate mean coverage in your textbook on page 160 (note that your task is easier because you only have to calculate the mean of column 4). You can select a single chromosome with ‘grep’ and calculate the mean coverage like this:

	$ cat genome_coverage.txt | grep –w 'chr1' | awk '{sum+=$3} END {print sum/NR}'

Reminder: what does the grep option '-w' do? 

Similarly, you can select a single chromosome and calculate a standard deviation like this: 

	$ cat genome_coverage.txt | grep –w 'chr1' | awk '{sum+=$3; sumsq+=$3*$3} END {print sqrt(sumsq/NR - (sum/NR)**2)}'

Try manipulating the different parts of the equation to figure out how awk is operating. If your genomecov script is taking a long time you can start to use these commands to experiment with calculations on your file. They will not be accurate because your file is not complete but you can use them on the partial file to make sure you can get them going. 

Remember that for diagnosis you must do this **for each chromosome separately** and use the z-score calculation we discussed in class to diagnose your sample. In order to fully present your results you must plot your data. For example, you could make a plot with chromosome on the x-axis and the average coverage +\- standard deviation on the y-axis. Keep in mind that I have been told numerous times I am “graphically challenged” and if you think you might have a more elegant way to plot and present the data, you are probably right!

Your final task is to write a short report (2-3 pages) summarizing your results. Here are some ideas for discussion points for your presentation and report:

1)	What is the difference between false positive and false negative results? Do you have concerns your results could be a false positive or negative?
2)	How might chromosome size or identity (autosome vs. sex chromosome, etc) affect your results? 
3)	How many sites in your sample have a genome sequence coverage > 20? What kind of coverage do you think is necessary for reliable results? 
4)	Does the standard deviation for coverage scale with the coverage level? For example, do you see smaller standard deviations with higher coverage levels? How does this vary across chromosomes?

Make sure you include plots that demonstrate the sequencing coverage and diagnosis (chromosomal abnormality or not). Your plots should include figure legends similar to what would be published in a scientific article. Make sure you label your axes and include any legends or other information that is necessary to understand your plot. A wise person once told me that a great figure is one where you don’t need to read the figure legend to understand it. Make that your goal. 

Your group report is due November 5, the first day after Fall Break 2019. 

