Today we will visualize some sequence reads that have been aligned to a reference genome. This tutorial will outline some of the sites and give you in-class questions to answer but we will work through these things as a group.

First, download the aligned sequences. I have placed them in the **shared** directory on ASC. Open your terminal and navigate to your desktop (it's easiest to find the files there on the lab computers). On MobaXterm this looks like:

    $ cd /home/Lab/Desktop

On a mac it looks like:

    $ cd ~/Desktop
    
Once you have navigated to your desktop transfer your files with scp (**secure copy**)

    $ scp ualcls*yournumberhere*@uv.asc.edu:~/shared/SL*bam*gz ./

These files are large and will take a little while to transfer (possibly >20 minutes depending on traffic).

While your files are transferring we will install and launch the Broad Integrated Genome Viewer, IGV. Navigate to this website http://software.broadinstitute.org/software/igv/download and launch the 1.2GB version. 

We will work together through loading your sequence read files into IGV and viewing them. For reference you will need to load the Dm3 version of the *Drosophila melanogaster* genome. 

Once your sequences have finished downloading unzip the files like this:

       $ gunzip *.gz

When you are visualizing them with IGV try to explore the program and features. Click around to zoom in and out, load different chromosomes, explore exon/intron boundaries and look at the polymorphisms between your samples and the reference sequence. 

For in-class points answer the following:

1) Choose 2 genes from your viewer (any 2, your choice, you pick). Go to FlyBase www.flybase.org and enter your gene into the search box. For each gene answer: 

  a) What is the gene name?
  b) What is the gene function?
  c) Does your gene show polymorphisms between your 2 samples? 
  d) Does your gene show polymorphisms between your sample and the reference genome? 
