Velvet/SPAdes assembly tutorial

The data for this week have been generously provided by Dr. Juan Lopez-Bautista’s lab. We will use the assembly program Velvet to assembly the sequences. Copy the data into your home directories on ASC. They are in the shared directory. Remember that in order to use them you need to copy them into your home directory and work there.

    $ cd shared
    $ cp *interleaved* ~

# This will copy everything with ‘fast’ in the name into your home directory.

First, do some sanity checks on the reads. How many reads are there? 

    $ wc –l G_interleaved.fastq

# What are the two files? How long are the reads?

# Velvet requires that paired reads are interleaved into a single file. 

First, load the Anaconda module 

        $ module load anaconda

Velvet should be accessible, test this and see if you get a program description

        $ velveth

# Then use the Velvet script shuffleSequences_fastq.pl to create an interleaved file.

#     $ /opt/asn/apps/velvet_1.2.10_131bp/bin/shuffleSequences_fastq.pl G_F.fastq G_R.fastq G_interleaved.fastq

# As a sanity check to make sure it worked, check the file

#     $ more G_interleaved.fastq

# and check the number of lines.

#     $ wc –l G_interleaved.fastq

The Velvet pipeline uses a two-step procedure for assembly. First, the sequences are preprocessed with the velveth program and then the assembly is run with velvetg. Type 

    $ velveth

to read the program options. 

We will need to write a script to run velvet. Inside your script, run velveth with an output directory and a k-mer value. Here, I have chosen 31 as the k-mer value and named my directory and output files with 31 so I can remember which value I chose. Use vi to write your script 

    $ vi velvet_script

Hit 'i' for insertion mode and type your script in

    #!/bin/bash

    source /opt/asn/etc/asn-bash-profiles/modules.sh

    module load velvet/1.2.10_131bp

    # place commands here

    velveth velvet_31 31 -fastq -short G_interleaved.fastq > velvet_31.velveth.stdOut

    velvetg velvet_31 -exp_cov auto -cov_cutoff auto

Save your script by typing ‘escape’ then :wq. Submit your script to the class queue

    $ run_script velvet_scipt

Once your reads are preprocessed (velveth) you will run the velvetg assembly portion. Here, I am running the assembly with the ‘expected coverage’ and ‘coverage cutoff’ values set automatically. 

Once your job has finished running you will see a directory named 'velvet\_31'. Navigate into this directory

    $ cd velvet_31

and examine your output files. 

The ‘Log’ file will tell you some statistics about your assembly including: 1) how many nodes you have; 2) N50; 3) maximum contiguous sequence size; 4) total assembly size; and 5) how many sequencing reads were used in creating your assembly. 

Ultimately one of the biggest determinants of assembly success is the k-mer value. Small k-mers can result in overly fragmented assemblies while too large k-mers will simply not work. Manipulate the k-mer value upwards and see how it affects your assembled genome sequence. You will need to run both velveth and velvetg again (edit your script for k-mer size). For example, for k=51 

        velveth velvet_51 51 -fastq -short G_interleaved.fastq > velvet_51.velveth.stdOut
        
        velvetg velvet_51 –exp_cov auto –cov_cutoff auto

If you experiment you will be able to get the entire dataset to assemble into one large contiguous sequence with a few small junk bits (hint: get aggressive about increasing the k-mer value). I have included an accessory script, count\_fasta.pl that can be used to produce a histogram-appropriate file of sequence lengths and some summary stats. You call it like this:

        $ perl count_fasta.pl –i 10000 ./velvet_51/contigs.fa

Where the ‘-i’ flag sets the bin size for the histogram and 

        ./velvet_51/contigs.fa 

is the path to your assembled contigs file. Try running it with –i 100 , -i 1000 and other values to see how your k-mer size affects the distribution of assembled sequences in your final output. 

Now we need to figure out what this organism is. The easiest way to do this when you have a large contiguous sequence is to copy a section of the assembled sequence and paste it into the NCBI Blast browser. 

Most modern de Bruijn graph assemblers use a similar framework but the actual implementation and results vary widely. In order to explore this we will also use the assembler SPAdes to assemble our genome sequence. Load the module with

        $ module load spades

This time you will need to submit a script that loads the module and calls the program with a script like the one below. Type

        $ vi spades_script.sh

Hit 'i' for insertion mode and type the following:

        #!/bin/bash

        source /opt/asn/etc/asn-bash-profiles/modules.sh

        module load spades

        spades.py -t 1 -m 12 -o pe --pe1-1 ./G_F.fastq --pe1-2 ./G_R.fastq | tee pe.stdOut

Hit 'escape' and ':wq' to save and exit. 

SPAdes is a more sophisticated program and does not require user-inputted parameters.

Your homework is to write a report describing the parameters you tried in your genome assembly and the influence of the parameters on your assembly. For example, how do smaller and larger k-mer values affect your velvet assemblies? Support your results with relevant graphs and plots. For example, you could plot your histogram output as a demonstration of assembly size distribution. How does the number of contiguous sequences vary with k-mer value? How does the assembly N50 vary with k-mer value? To address these questions you could present graphs relating k-mer size with N50, number of contiguous sequences in the assembly, or other statistic. In order to get full credit you should try at least 5 k-mer values.

You must also answer these specific questions:
1.	What is your organism?
2.	What kind of DNA did you assemble? 
3.	What are the final statistics for your best Velvet and SPAdes assemblies (Total size, N50, number of scaffolds)?

You must print your report and hand it in. You may either drop it by my office (SEC 2328) or the Biological Sciences office (SEC 1325). 
